op-challenger/game/fault/types/types.go

```diff
@@ -166,3 +167,46 @@ type Claim struct {
 func (c *Claim) IsRoot() bool {
 	return c.Position.IsRootPosition()
 }
+
+// ChessTime returns the amount of time accumulated in the chess clock.
+// Does not assume the claim is countered and uses the specified time
+// to calculate the time since the claim was posted.
+func (c *Claim) ChessTime(now time.Time) time.Duration {
+	timeSince := int64(0)
+	if now.Unix() > int64(c.Clock.Timestamp) {
+		timeSince = now.Unix() - int64(c.Clock.Timestamp)
+	}
+	return time.Duration(c.Clock.Duration) + time.Duration(timeSince)
+}
+
+// Clock is the clock type used in the FaultDisputeGame contract.
+// It is a packed uint128 with the upper 64 bits being the
+// duration and the lower 64 bits being the timestamp.
+// ┌────────────┬────────────────┐
+// │    Bits    │     Value      │
+// ├────────────┼────────────────┤
+// │ [0, 64)    │ Duration       │
+// │ [64, 128)  │ Timestamp      │
+// └────────────┴────────────────┘
+type Clock struct {
+	Duration  uint64
+	Timestamp uint64
+}
+
+// NewClock creates a new Clock instance.
+func NewClock(raw *big.Int) *Clock {
+	maxUint64 := new(big.Int).Add(new(big.Int).SetUint64(math.MaxUint64), big.NewInt(1))
+	remainder := new(big.Int)
+	quotient, _ := new(big.Int).QuoRem(raw, maxUint64, remainder)
+	return &Clock{
+		Duration:  quotient.Uint64(),
+		Timestamp: remainder.Uint64(),
+	}
+}
+
+// Packed returns the Clock as a packed uint128.
+func (c *Clock) Packed() *big.Int {
+	duration := new(big.Int).SetUint64(c.Duration)
+	encoded := new(big.Int).Lsh(duration, 64)
+	return new(big.Int).Or(encoded, new(big.Int).SetUint64(c.Timestamp))
+}
```

Comment:
This is leaking contract representation of the data outside of the contracts package. We should have this common type just model based on the duration and timestamp being separate and move this encode/decode logic into the contracts package.

I know it's always an appealing design to have an object that knows how to encode and decode itself but that only works when the encoding is somehow intrinsic to the type. It breaks down once you start to have multiple encodings of a type, like when we need to support multiple different contract versions. The idea of the custom contract bindings is that they will provide an anticorruption layer and prevent the specific details of the contract implementations from "leaking" into the challenger codebase.

Reply:
Personally, since these are explicitly `fault` game types that correspond to the FaultDisputeGame contract (as mentioned across the fault types source [example here](https://github.com/ethereum-optimism/optimism/blob/develop/op-challenger/game/fault/types/types.go)), I think it would make more sense to tie sepecific types to their game types + contract implementations so the logic is consistent with that of the contract.

Additionally, in this case I do believe the encoding/decoding _is_ intrinsic to the type. The `Clock` object is a explicitly defined as a packed duration and timestamp and that is managed in it's type library in solidity. There can't be other encodings of this type as I understand it.

We can change this pr up to perform the decoding in the contracts package but it will bubble up the stack and take a bit of time. Do you think this type leakage is not acceptable? I'd say it's not prohibitive because the type is isolated in solidity and it's tested here.

Reply:
I'm going to be quite stubborn here sorry. :) The point of the contracts package is to act as an anti-corruption layer and keep the challenger and contracts as two separate bounded contexts.  Being separate bounded contexts, they can and and do have different architectures and model concepts in different ways. The chess clock is actually a good example of that - it's the same concept in both, but in the contracts it's represented as a single packed uint256 whereas in the challenger, even in this PR it's represented as two separate values. Similar differences exist for how we model claims, games and many other concepts.  The contracts package is responsible for doing any required translation between the two bounded contexts. Even more strongly than that, part of the responsibility of the contacts package is to prevent the way that the contracts represent things from leaking through and corrupting the way the challenger represents things (and visa versa).  I have a sermon on this topic if you still aren't convinced: https://www.notion.so/oplabs/Facades-and-Adapters-and-Anti-corruption-Layers-Oh-My-4b522f20100f419bbea8123ae21a6b53?pvs=4 

In terms of different implementations, we've already had a period where the contracts were under-going significant changes and we supported two versions of them for a while. It was this anti-corruption layer that let us do that and make incremental progress rather than having to either have one gigantic PR or leave tests broken for a long period.